### PR TITLE
Map first and last tokens in original_range

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR try to fix the first part of the `original_range` : Try to map the first token and last token of a `SyntaxNode` , If success, return the union range of mapped tokens.

### CODE REVIEW

File: crates/ra_ide/src/expand.rs [hunk 1]
```diff
@@ -17,26 +17,45 @@ pub(crate) fn original_range(db: &RootDatabase, node: InFile<&SyntaxNode>) -> Fi
...
+                // FIXME: Is is possible ?
+                if first.file_id != last.file_id {
+                    return None;
+                }
+
+                // FIXME: Add union method in TextRange
+                let range = union_range(first.value.text_range(), last.value.text_range());
+                Some(first.with_value(range))
+            }
         }
```
Comment (matklad) [inline crates/ra_ide/src/expand.rs]: I think there's a much better way to write this, but let's merge this to unblock the progress of tokens with ids, and refactor afterwards!

**Comment by edwin0cheng** [comment]
I think we can't only check the first and last tokens only, For example: 
```rust
macro_rules! define_fn {
    ($name:ident) => (fn $name() {})  // <- **A**
}

define_fn!(
    foo       // <- **B**
)

fn bar() {
   <|>foo();
}
```

For goto definition, we want to go to **B** instead of **A**, however, note that the full-range of the `foo()` will point to A, because `fn` and `{}` is inside **A**. To fix this problem, we have two solutions:

1. Check all tokens instead of only first and last tokens.
2. Let the full range point to **A**, but we want to change the definition of `NavigationTarget` to:
```rust
pub struct NavigationTarget {
    ...
    full_range: (FileId, TextRange),
    focus_range: Option<(FileId, TextRange)>,
}

Personally I prefer the first option.

**Comment by matklad** [comment]
Oh, that's an interesting example! Basically, goto definition should return two ranges: `focusRange`, where the caret is positioned, and `fullRange`, encompassing the whole node.

In the example above, the `focusRange` is inside a macro call, but fullRange is inside expansion! That is, the natural mapping uses two different macro files foe them.

I think this means that the precise mapping here is really the concern of `NavigationTarget`. It should try to (in order of preference) 

1. map both ranges to macro_call
2. map focus range to a token inside macro call, and full range to the whole of macro call
3. map both ranges to the whole of macro call

**Comment by edwin0cheng** [comment]
One more question, what is the `focus_range` and `full_range` of `foo` in the following example ?

```rust
macro_rules! define_fn {
    () => (fn foo() {})  // <- **A**
}

define_fn!();  // <- **B**

fn bar() {
   <|>foo();
}
```

**Comment by matklad** [comment]
I hint both should be equal to `define_fn!(‘` invocation. Ie, we never should jump into macro definition

**Pull request merged**